applications/nrf5340_audio/src/audio/streamctrl_unicast_server.c

```diff
@@ -0,0 +1,722 @@
+							K_NO_WAIT);
+		ERR_CHK(ret);
+
+		data_fifo_block_free(&ble_fifo_rx, &stale_data);
+	}
+
+	ret = data_fifo_pointer_first_vacant_get(&ble_fifo_rx, (void *)&iso_received, K_NO_WAIT);
+	ERR_CHK_MSG(ret, "Unable to get FIFO pointer");
+
+	if (data_size > ARRAY_SIZE(iso_received->data)) {
+		ERR_CHK_MSG(-ENOMEM, "Data size too large for buffer");
+		return;
+	}
+
+	memcpy(iso_received->data, p_data, data_size);
+
+	iso_received->bad_frame = bad_frame;
+	iso_received->data_size = data_size;
+	iso_received->sdu_ref = sdu_ref;
+	iso_received->recv_frame_ts = recv_frame_ts;
+
+	ret = data_fifo_block_lock(&ble_fifo_rx, (void *)&iso_received,
+				   sizeof(struct ble_iso_data));
+	ERR_CHK_MSG(ret, "Failed to lock block");
+}
+
+/* Thread to receive data from BLE through a k_fifo and send to audio datapath */
+static void audio_datapath_thread(void *dummy1, void *dummy2, void *dummy3)
+{
+	int ret;
+	struct ble_iso_data *iso_received = NULL;
+	size_t iso_received_size;
+
+	while (1) {
+		ret = data_fifo_pointer_last_filled_get(&ble_fifo_rx, (void *)&iso_received,
+							&iso_received_size, K_FOREVER);
+		ERR_CHK(ret);
+
+		audio_datapath_stream_out(iso_received->data, iso_received->data_size,
+					  iso_received->sdu_ref, iso_received->bad_frame,
+					  iso_received->recv_frame_ts);
+		data_fifo_block_free(&ble_fifo_rx, (void *)&iso_received);
+
+		STACK_USAGE_PRINT("audio_datapath_thread", &audio_datapath_thread_data);
+	}
+}
+
+/* Function for handling all stream state changes */
+static void stream_state_set(enum stream_state stream_state_new)
+{
+	strm_state = stream_state_new;
+}
+
+uint8_t stream_state_get(void)
+{
+	return strm_state;
+}
+
+void streamctrl_send(void const *const data, size_t size, uint8_t num_ch)
+{
+	int ret;
+	static int prev_ret;
+
+	struct encoded_audio enc_audio = {.data = data, .size = size, .num_ch = num_ch};
+
+	if (strm_state == STATE_STREAMING) {
+		ret = unicast_server_send(enc_audio);
+
+		if (ret != 0 && ret != prev_ret) {
+			if (ret == -ECANCELED) {
+				LOG_WRN("Sending operation cancelled");
+			} else {
+				LOG_WRN("Problem with sending LE audio data, ret: %d", ret);
+			}
+		}
+
+		prev_ret = ret;
+	}
+}
+
+#define TEST_TONE_BASE_FREQ_HZ 1000
+
+static int test_tone_start(void)
+{
+	int ret;
+	static uint32_t test_tone_hz;
+
+	if (CONFIG_AUDIO_BIT_DEPTH_BITS != 16) {
+		LOG_WRN("Tone gen only supports 16 bits");
+		return -ECANCELED;
+	}
+
+	if (strm_state == STATE_STREAMING) {
+		if (test_tone_hz == 0) {
+			test_tone_hz = TEST_TONE_BASE_FREQ_HZ;
+		} else if (test_tone_hz >= TEST_TONE_BASE_FREQ_HZ * 4) {
+			test_tone_hz = 0;
+		} else {
+			test_tone_hz = test_tone_hz * 2;
+		}
+
+		if (test_tone_hz != 0) {
+			LOG_INF("Test tone set at %d Hz", test_tone_hz);
+		} else {
+			LOG_INF("Test tone off");
+		}
+
+		ret = audio_encode_test_tone_set(test_tone_hz);
+		ERR_CHK_MSG(ret, "Failed to generate test tone");
+	}
+
+	return 0;
+}
+
+/* Handle button activity */
```

Comment:
For concistency, add @note for these zbus threads documentations, see e.g. streamctrl_broadcast_source for reference